### DISCUSSION TITLE
Maybe the convergence method should be improved?

### DISCUSSION THREAD

**Discussion by ceraolo** [OpenModelica/OpenModelica#10321] [General]
Using variable-step algorithms should allow users to forget about carefully choosing the timestep, since it is determined by the algorithm itself.
Consider models Equi and NoEqui included in the following package:
[Test.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10896579/Test.mo.txt)

Equi  contains equidistant time step and requests 3000 steps. NoEqui  does not request equidistant time grid. One would expect the plot results equal to each other, or, at most, some lesser detail in case of equidistant time step in comparison with the NoEqui ouput.
That's why the following plots are very surprising:
![image](https://user-images.githubusercontent.com/7373030/223072991-dbdadb5a-335a-4f0f-a763-d47e9ce0fbb2.png)

As you can see from is NoEqui, dassl jumps over sine waves, disregarding many details of the current shape (correctly shown in Equi).  Nevertheless, the machine torque in NoEqui is ok.

This hints that the criterion to determine convergence, and therefore the time step, does not consider all the output variables (nor all the state variables, since currents, I think, are state variables in this model), and this can produce very large errors. I got similar behaviour also with ida and cvode.

Is this unavoidable? or it can be fixed adding (possibly optionally) an additional convergence criterion, taking into account more output variables (e.g. all state variables)?

Note that Dymola shows similar behaviour.

**Comment by casella** [comment]
@ceraolo, this is a good question. As you mentioned, ODE variable step-size solvers only care about the precision of state variables, algebraic variables are only computed to eventually get the state derivatives, but there is no error control on them. This is common to all Modelica tools that turn DAEs into ODEs and then use ODE solvers.

I think what you are experiencing here is simply an [aliasing](https://en.wikipedia.org/wiki/Aliasing) issue. The Park transformations are meant precisely to avoid the solver the effort of tracking variables that are oscillating like crazy, since direct and quadrature currents change much more slowly than actual currents and voltages at the machine port. The goal of the ODE solver at each time step is to compute the state vector at the next time step with an estimated error below the selected tolerance; it doesn't care what happens in-between. Since ODE integration is a computationally intensive task, particularly for implicit solvers such as DASSL or IDA, this is absolutely reasonable: there is no point taking smaller steps to predict the next states with enough accuracy. Then, of course, you still have to compute the outputs as a function of the interpolated states, using the output equations.

Solvers with so-called dense output (including DASSL and IDA) also provide the information to interpolate the state values for intermediate times in-between integration steps, with a precision that has the same order of magnitude as the selected tolerance - typically they use polynomials of high enough order, that approximate the state trajectory. The idea is that this interpolation is way cheaper than the computation of a new step of the integration algorithm.

Equidistant time grid output uses this dense output to interpolate the state values on a regular grid, and then compute all the outputs using the output equations, which is also in general less computationally expensive than computing a step with an implicit solver. However, they do it for the points you are asking, i.e., on a regularly spaced grid. It is up to you to decide how close to each other should be the grid points, the solver has no idea. 

Note that, albeit the plot on the top left seems very bad, the individual points that make it are correct - it's just the straight lines in-between them that do not correspond to the real system behaviour. Again, the solver has no clue.

If you don't want to set the equidistant time grid yourself, a dirty trick is always possible, i.e. to add one or more dummy differential equation `der(dummy) = y`, where y is the output that you want to see, and then disable the equidistant time grid output. This forces the ODE solver to take much smaller steps to follow the fast variations of the outputs. However, this may be quite inefficient, because then each saved point requires a full evaluation of the implicit ODE solver to compute the state values, rather than a simple dense output interpolation. It may be advantageous if your output change a lot in some intervals and are nearly constant in others, because then you save a lot of steps in the latter case, which would be unnecessarily computed on the equidistant time grid, and this could actually compensate for the extra time taken for the step computations that is not using the dense output interpolation.

Which is the best solution also depends on the ratio between two amounts of time: the amount taken to compute the ODE RHS and to solve the implicit integration formula, vs the time needed to interpolate the states and compute the output equations. If the former dominates, dense output and equidistant time grid is better; if the latter dominates, adding the dummy output differential equation may be beneficial.

Adding @bernhardbachmann to the loop, if he wants to comment more.

**Reply by bernhardbachmann** [reply]
Nothing to add to that! You have explained the phenomenon very thoroughly

**Comment by ceraolo** [comment]
Thank you, @casella,
I think you've analysed this thing in depth, and clarified everything.
I was misleaded by the fact that currents usually are state variables, but this is not true here. Here, the machine uses internally park transformation, and therefore in this case the state variables become Id and Iq, which vary much slower than I1, I2, i3.
Thanks again, very good analysis.

**Comment by casella** [comment]
In fact, that's the whole point of using Park's transformation 😃
